File: android/src/main/java/com/lightningkite/kwift/observables/TransformedEvent.shared.kt
package com.lightningkite.kwift.observables

import com.lightningkite.kwift.escaping

class TransformedEvent<A, B>(
    val basedOn: Event<A>,
    val transformation: @escaping() (A) -> B
) : Event<B>() {
    override fun add(listener: @escaping() (B) -> Boolean): Close {
        val transformation = this.transformation
        return basedOn.add { it ->
            listener(transformation(it))
        }
    }
}

@Deprecated("Use 'map' instead", ReplaceWith(
    "this.map(transformation)",
    "com.lightningkite.kwift.observables.map"
)
)
fun <T, B> Event<T>.transformed(transformation: @escaping() (T) -> B): Event<B> {
    return TransformedEvent<T, B>(this, transformation)
}

fun <T, B> Event<T>.map(transformation: @escaping() (T) -> B): Event<B> {
    return TransformedEvent<T, B>(this, transformation)
}

File: ios/Kwift/com/lightningkite/kwift/observables/TransformedEvent.shared.swift
//Package: com.lightningkite.kwift.observables
//Converted using Kwift2

import Foundation



public class TransformedEvent<A, B>: Event<B> {
    
    public var basedOn: Event<A>
    public var transformation:  (A) -> B
    
    
    override public func add(listener: @escaping (B) -> Bool) -> Close {
        var transformation = self.transformation
        return basedOn.add{ (it) in 
            listener(transformation(it))
        }
    }
    
    public init(basedOn: Event<A>, transformation: @escaping (A) -> B) {
        self.basedOn = basedOn
        self.transformation = transformation
        super.init()
    }
    convenience public init(_ basedOn: Event<A>, _ transformation: @escaping (A) -> B) {
        self.init(basedOn: basedOn, transformation: transformation)
    }
}
 

extension Event {
     public func transformed<B>(transformation: @escaping (T) -> B) -> Event<B> {
        return TransformedEvent<T, B>(self, transformation)
    }
}
 
 

extension Event {
    public func map<B>(transformation: @escaping (T) -> B) -> Event<B> {
        return TransformedEvent<T, B>(self, transformation)
    }
}
 
